File: app/src/main/java/com/lbynet/phokus/utils/MathTools.java
package com.lbynet.phokus.utils;

import android.util.SizeF;

public class MathTools {

    final public static float fullFrameCmosSize = (float)Math.sqrt(Math.pow(36,2) + Math.pow(24,2));

    public static float getCappedFloat(float value, float min, float max) {

        if(min > max) {
            float temp = min;
            min = max;
            max = temp;
        }

        if(value > max) { return max; }

        else if (value < min) { return min; }

        else { return value; }
    }

    public static boolean isValueInRange(int value, int low, int high) {
        return low > high ?
                    (value >= high && value <= low)
                :   (value >= low && value <= high);
    }

    public static float getCropFactor(SizeF cmosDimensions) {
        return fullFrameCmosSize / (float)(Math.sqrt(Math.pow(cmosDimensions.getWidth(),2) + Math.pow(cmosDimensions.getWidth() /3 * 2,2)));
    }

    public static int sumOf(int... values) {
        int sum = 0;

        for(int i : values) {
            sum += i;
        }

        return sum;
    }

    public static int [] sliceTime(double ms) {

        return new int [] {
                (int)Math.floor(ms / 1000 / 3600), //Hour
                (int)Math.floor((ms / 1000 /60) % 60), //Minute
                (int)Math.floor((ms / 1000) % 60), //Second
                (int)Math.floor(ms % 100) //Tens of ms
        };

    }

    /**
     * Caps radian to [0,2\pi].
     * @param rad raw radian [-\infinity,+\infinity].
     * @return capped radian.
     */
    public static double getCappedRadian(double rad) {

        final double twoPi = 2 * Math.PI;
        double remainder = rad % (twoPi);

        if(remainder < 0) remainder += twoPi;

        return remainder;
    }

    public static double radianToDegrees(double rad, boolean isAlwaysPositive) {

        if(!isAlwaysPositive) return radianToDegrees(rad);
        else return radianToDegrees(getCappedRadian(rad));

    }

    public static double radianToDegrees(double rad) {
        return rad * 180 / Math.PI;
    }

    public static void formatTime(double ms, String [] buffer) {
        int [] time = sliceTime(ms);

        if(buffer.length < 4) return;

        for(int i = 0; i < 4; ++i) {
            if (time[i] < 10) {
                buffer[i] = "0" + time[i];
            } else buffer[i] = Integer.toString(time[i]);
        }
    }
}

File: app/src/main/java/com/lbynet/phokus/utils/MathTools.kt
package com.lbynet.phokus.utils

import android.util.SizeF
import com.lbynet.phokus.utils.MathTools
import java.lang.StringBuilder

object MathTools {

    private val fullFrameCmosSize = Math.sqrt(Math.pow(36.0, 2.0) + Math.pow(24.0, 2.0))
        .toFloat()

    @JvmStatic
    fun getCappedFloat(value: Float, min: Float, max: Float): Float {
        var min = min
        var max = max
        if (min > max) {
            val temp = min
            min = max
            max = temp
        }
        return if (value > max) {
            max
        } else if (value < min) {
            min
        } else {
            value
        }
    }

    @JvmStatic
    fun isValueInRange(value: Int, low: Int, high: Int): Boolean {
        return if (low > high) value in high..low else value in low..high
    }

    @JvmStatic
    fun capValue(value : Float, low : Float, high : Float) : Float {

        if(value < low) return low
        else if(value > high) return high
        else return value
    }

    @JvmStatic
    fun getCropFactor(cmosDimensions: SizeF): Float {
        return fullFrameCmosSize / Math.sqrt(
            Math.pow(
                cmosDimensions.width.toDouble(),
                2.0
            ) + Math.pow((cmosDimensions.width / 3 * 2).toDouble(), 2.0)
        ).toFloat()
    }

    @JvmStatic
    fun sumOf(vararg values: Int): Int {
        var sum = 0
        for (i in values) {
            sum += i
        }
        return sum
    }

    @JvmStatic
    fun sliceTime(ms: Double): IntArray {
        return intArrayOf(
            Math.floor(ms / 1000 / 3600).toInt(),  //Hour
            Math.floor(ms / 1000 / 60 % 60).toInt(),  //Minute
            Math.floor(ms / 1000 % 60).toInt(),  //Second
            Math.floor(ms % 100).toInt() //Tens of ms
        )
    }

    /**
     * Caps radian to [0,2\pi].
     * @param rad raw radian [-\infinity,+\infinity].
     * @return capped radian.
     */
    @JvmStatic
    fun getCappedRadian(rad: Double): Double {
        val twoPi = 2 * Math.PI
        var remainder = rad % twoPi
        if (remainder < 0) remainder += twoPi
        return remainder
    }

    @JvmStatic
    fun radianToDegrees(rad: Double, isAlwaysPositive: Boolean): Double {
        return if (!isAlwaysPositive) radianToDegrees(rad) else radianToDegrees(
            getCappedRadian(rad)
        )
    }

    @JvmStatic
    fun radianToDegrees(rad: Double): Double {
        return rad * 180 / Math.PI
    }

    @JvmStatic
    fun formatTime(ms: Double, buffer: Array<String?>) {
        val time = sliceTime(ms)
        if (buffer.size < 4) return
        for (i in 0..3) {
            if (time[i] < 10) {
                buffer[i] = "0" + time[i]
            } else buffer[i] = Integer.toString(time[i])
        }
    }

    @JvmStatic
    fun parseTimeToString(elaspedTimeInMs: Double): String {
        val buffer = arrayOfNulls<String>(4)
        formatTime(elaspedTimeInMs, buffer)
        val sb = StringBuilder().append(buffer[0])
        for (i in 1..3) {
            sb.append(':').append(buffer[i])
        }
        return sb.toString()
    }
}